### PR TITLE
Add rotation support and other graphics fixes

### PR DESCRIPTION
**Pull request by iRath96**
Hi there 👋 
Thank you for this awesome repository 👍 

I've worked on a branch that addresses the following issues:
* Model rotations (fixes issue #34)
* Handling transparent blocks better (fixes issue #33)
* Implement multi-part block models (fixes issue #23)
* Respects pixel device ratios (beneficial for HiDPI displays)
* Fixed UV coordinates for blocks like hoppers
* Liquids should be 14px instead of 16px high

Here a before…
Bildschirmfoto 2020-10-16 um 00 04 58

…and after:
Bildschirmfoto 2020-10-16 um 00 00 20

Face culling (issue #33) might still not be perfect, but is significantly improved. Also note that I had to move model rotation from the server to the client in order for element rotations to work correctly – for me this caused processing times to go up from 400ms to 800ms per section. If this is a concern to you then perhaps implementing a per-block-variant cache for storing vertex positions and other attributes would be an option to bring performance back to (or even improve upon) previous versions.

Please give me some feedback and let me know what needs to be changed for this to get merged. Thank you! 👏

### CODE REVIEW

**Comment by Karang** [comment]
This looks very good. I'll try to review this as soon as possible.

Can you run the linter to fix the failing tests ? `npm run fix`

**Comment by Karang** [comment]
Thanks for implementing block rotations and multipart blocks, this part looks definitely better ;)

But there are some issues with your face culling implementation:

It seems that you create a face (potentially 2) between all blocks that are different, but this create a lot of non-visible faces that can affect performances.

Before:
image

After:
image

The leaves, even if they are the same blocks, do have faces between them:

Before:
image

After:
image

File: src/models.js [hunk 1]
```diff
@@ -108,22 +113,68 @@ function renderLiquid (world, cursor, texture, type, water, attr) {
...
+}
+
+function renderElement (world, cursor, element, doAO, attr, globalMatrix, globalShift, block) {
   for (const face in element.faces) {
     const eFace = element.faces[face]
     const { dir, corners, mask1, mask2 } = elemFaces[face]
 
     if (eFace.cullface) {
       const neighbor = world.getBlock(cursor.plus(dir))
-      if (!neighbor || !(neighbor.transparent || !neighbor.isCube) || neighbor.position.y < 0) continue
```
Comment (Karang) [inline src/models.js:118]: If you want to address the culling issue later, you can revert this change so we can merge already the other improvements.

**Review by Karang — CHANGES_REQUESTED** [review]

File: src/models.js (hunk 1 above)
Comment (iRath96) [reply src/models.js:118]: Thanks for letting me know. I have fixed this in d502c90a740cbad3e476896b086da71c327f5919. Apparently this was also the reason the section processing time went up, which is now back to what it was before this PR (i.e. doing the rotations on the client does not cause any noticeable overhead).

**Comment by Karang** [comment]
It is much better now, thanks for the fix.

There is still a smaller issue again related to the culling:
image
It seems to happen on certain rotations of logs and stairs.

**Comment by iRath96** [comment]
My bad. I completely forgot that the direction vectors for culling and lighting need to be rotated as well. Now it should be working.

File: src/models.js [hunk 2]
```diff
@@ -70,12 +70,17 @@ const elemFaces = {
 }
 
 function renderLiquid (world, cursor, texture, type, water, attr) {
+  const blockAbove = world.getBlock(cursor.plus(new Vec3(0, 1, 0)))
+  const liquidHeight = (blockAbove.type === type ? 16 : 14) / 16
+
   for (const face in elemFaces) {
     const { dir, corners } = elemFaces[face]
 
     const neighbor = world.getBlock(cursor.plus(dir))
```
Comment (Karang) [inline src/models.js:79]: dir is not a vec3 anymore with your latest change, so plus() will fail

**Review by Karang — CHANGES_REQUESTED** [review]

File: src/models.js (hunk 2 above)
Comment (iRath96) [reply src/models.js]: Fixed in 241895917ee18d646f16aa77ee224a8f425d9bc1.

**Pull request merged**